extension/js/common/oauth2/oauth2.ts

```diff
@@ -0,0 +1,52 @@
+/* ©️ 2016 - present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com */
+/* eslint-disable @typescript-eslint/no-explicit-any  */
+import { Bm, BrowserMsg } from '../browser/browser-msg.js';
+import { windowsCreate } from '../browser/chrome.js';
+
+export class OAuth2 {
+
+  public static webAuthFlow = async (url: string): Promise<Bm.AuthWindowResult> => {
+    const screenWidth = (window.screen.width || window.innerWidth);
+    const screenHeight = (window.screen.height || window.innerHeight);
+    // non-standard but supported by most of the browsers
+    const { availLeft, availTop } = (window.screen as unknown as { availLeft?: number, availTop?: number });
+    let adaptiveWidth = Math.floor(screenWidth * 0.4);
+    if (adaptiveWidth < 550) {
+      adaptiveWidth = Math.min(550, Math.floor(screenWidth * 0.9));
+    }
+    const adaptiveHeight = Math.floor(screenHeight * 0.9);
+    const leftOffset = Math.floor((screenWidth / 2) - (adaptiveWidth / 2) + (availLeft || 0));
+    const topOffset = Math.floor((screenHeight / 2) - (adaptiveHeight / 2) + (availTop || 0));
+    const oauthWin = await windowsCreate({ url, left: leftOffset, top: topOffset, height: adaptiveHeight, width: adaptiveWidth, type: 'popup' });
+    if (!oauthWin || !oauthWin.tabs || !oauthWin.tabs.length || !oauthWin.id) {
+      return { error: 'No oauth window returned after initiating it' };
+    }
+    const tabId = oauthWin?.tabs && oauthWin.tabs[0].id;
+    return await new Promise((resolve) => {
+      BrowserMsg.addListener('auth_window_result', async (result: Bm.AuthWindowResult) => {
+        resolve(result);
```

Comment:
I meant something like this. Resolve the promise when the message is received. Wouldn't this work?

Reply:
Ohh.. Yeah.
It looks great. It's exactly what I wanted to achieve.
Thank you

Reply:
This method has problem though which will only accept `auth_window_result` event once.
Like when user wants to change email in `Force Google Account email change`(in experiemental page), event doesn't get fired.
Finding solution now.

Reply:
Yes, we need to remove both listeners when resolving the promise

Reply:
I thought experiemental `Force Google Account email change` feature doesn't work because event only fires once but that seems wrong.
It seems like event doesn't get fired in experimental page because it's wrapped in Iframe.

Reply:
@tomholub , @rrrooommmaaa Does anyone know why we use `BrowserMsg.send.bg.await.reconnectAcctAuthPopup` instead of using `GoogleAuth.newAuthPopup` directly?
Seems like UI test (enable google contact search test) fails because of this reason.

https://github.com/FlowCrypt/flowcrypt-browser/blob/37d466bf471bde88eae19a2ccd17f1bf74ce87d2/extension/chrome/elements/compose-modules/compose-recipients-module.ts#L725

Reply:
I don't see any good reason. Maybe some reason used to exist, but we refactored it away and it no longer makes sense. Feel free to remove `reconnectAcctAuthPopup`

Reply:
ok